### PR TITLE
Added Triangular type, At and Set methods, and tests

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

File: mat64/triangular.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+var (
+	triangular *Triangular
+
+	_ Matrix  = triangular
```
Comment (sbinet) [inline mat64/triangular.go]: naive comment: isn't "cleaner" to just spell it out
`_ Matrix = (*Triangular)(nil)`

admittedly more verbose, but no "pollution" of the global scope.

File: mat64/index_bound_checks.go [hunk 2]
```diff
@@ -28,10 +30,54 @@ func (m *Dense) Set(r, c int, v float64) {
...
+func (t *Triangular) set(r, c int, v float64) {
+	if r >= t.mat.N || r < 0 {
+		panic(ErrRowAccess)
+	}
+	if c >= t.mat.N || c < 0 {
+		panic(ErrColAccess)
+	}
+	if t.mat.Uplo == blas.Upper && r > c {
+		panic("mat64: triangular set out of bounds")
+	} else if t.mat.Uplo == blas.Lower && r < c {
```
Comment (fhs) [inline mat64/index_bound_checks.go]: Remove "else". It's idiomatic to write simple if statements. See [Effective Go](https://golang.org/doc/effective_go.html#if).

File: mat64/index_no_bound_checks.go [hunk 3]
```diff
@@ -35,3 +37,47 @@ func (m *Dense) Set(r, c int, v float64) {
...
+func (t *Triangular) Set(r, c int, v float64) {
+	if r >= t.mat.N || r < 0 {
+		panic(ErrRowAccess)
+	}
+	if c >= t.mat.N || c < 0 {
+		panic(ErrColAccess)
+	}
+	if t.mat.Uplo == blas.Upper && r > c {
+		panic("mat64: triangular set out of bounds")
+	} else if t.mat.Uplo == blas.Lower && r < c {
```
Comment (fhs) [inline mat64/index_no_bound_checks.go]: ditto

File: mat64/index_bound_checks.go [hunk 4]
```diff
@@ -28,10 +30,54 @@ func (m *Dense) Set(r, c int, v float64) {
...
+		}
+		return t.mat.Data[r*t.mat.Stride+c]
+	}
+	if r < c {
+		return 0
+	}
+	return t.mat.Data[r*t.mat.Stride+c]
+}
+
+// Set sets the element at row r and column c. Set panics if the
```
Comment (vladimir-ch) [inline mat64/index_bound_checks.go]: Unfinished comment

File: mat64/index_no_bound_checks.go [hunk 5]
```diff
@@ -35,3 +37,47 @@ func (m *Dense) Set(r, c int, v float64) {
...
+		}
+		return t.mat.Data[r*t.mat.Stride+c]
+	}
+	if r < c {
+		return 0
+	}
+	return t.mat.Data[r*t.mat.Stride+c]
+}
+
+// Set sets the element at row r and column c. Set panics if the
```
Comment (vladimir-ch) [inline mat64/index_no_bound_checks.go]: Unfinished comment

File: mat64/triangular.go (hunk 1 above)
Comment (btracey) [inline mat64/triangular.go]: I'm following the pattern that exists there now. We can change it if there's a better way, but consistency is better.

File: mat64/index_no_bound_checks.go (hunk 3 above)
Comment (btracey) [inline mat64/index_no_bound_checks.go]: Done.

File: mat64/index_bound_checks.go (hunk 2 above)
Comment (btracey) [inline mat64/index_bound_checks.go]: Done.

File: mat64/index_bound_checks.go (hunk 4 above)
Comment (btracey) [inline mat64/index_bound_checks.go]: Done.

File: mat64/index_no_bound_checks.go (hunk 5 above)
Comment (btracey) [inline mat64/index_no_bound_checks.go]: Done.

**Comment by btracey** [comment]
Fixed comments and added a comment to NewTriangular about the underlying data. PTAL

**Comment by jonlawlor** [comment]
Is QR going to be changed to:
```go
type QRFactor struct {
	QR    *Triangular
	rDiag []float64
}
```

**Comment by btracey** [comment]
I would vote it gets changed in some form. You can see my thoughts on Eigen at Issue #79 , and I have similar thoughts for Cholesky and QRFactor. I'm not a dictator though, we'll need to discuss how we want these to look. Getting the basic types in will make it easier to think about.

**Comment by btracey** [comment]
We should also consider if we want the Upper and Lower constants, or if we would rather have a boolean where true == Upper. There are only two possibilities, it would eliminate a panic possibility, and a boolean would match what we've chosen for Trans. However, it seemed to be that Upper and Lower read better, and it's not clear a priori which one should be true.

**Comment by jonlawlor** [comment]
You can type Upper and Lower to boolean and callers can just use true and false.  http://play.golang.org/p/9FKN8h4SbT

**Comment by btracey** [comment]
Removed TriType and made the field a Boolean. We can revisit if we want to create a typed Boolean, but this is in line with the current Trans behavior.

**Comment by jonlawlor** [comment]
That bool change looks good to me; but I haven't reviewed all of the other changes in depth.

**Comment by btracey** [comment]
Pulled from master

**Comment by btracey** [comment]
Just made a new PR ( #94) that incorporates some changes from the symmetric discussion.

**Pull request closed without merging**